src/Build/Logging/BinaryLogger/BinaryLogger.cs

```diff
@@ -0,0 +1,136 @@
+﻿using System;
+using System.IO;
+using System.IO.Compression;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Logging
+{
+    /// <summary>
+    /// A logger that serializes all incoming BuildEventArgs in a compressed binary file (*.binlog). The file
+    /// can later be played back and piped into other loggers (file, console, etc) to reconstruct the log contents
+    /// as if a real build was happening. Additionally, this format can be read by tools for
+    /// analysis or visualization. Since the file format preserves structure, tools don't have to parse
+    /// text logs that erase a lot of useful information.
+    /// </summary>
+    public class BinaryLogger : ILogger
+    {
+        private const int FileFormatVersion = 1;
+
+        private Stream stream;
+        private BinaryWriter binaryWriter;
+        private BuildEventArgsWriter eventArgsWriter;
+
+        private string FilePath { get; set; }
+
+        /// <summary>
+        /// The binary logger Verbosity is always maximum (Diagnostic). It tries to capture as much
+        /// information as possible.
+        /// </summary>
+        public LoggerVerbosity Verbosity { get; set; } = LoggerVerbosity.Diagnostic;
+
+        /// <summary>
+        /// The only supported parameter is the output log file path (e.g. "msbuild.binlog") 
+        /// </summary>
+        public string Parameters { get; set; }
+
+        /// <summary>
+        /// Initializes the logger by subscribing to events of IEventSource
+        /// </summary>
+        public void Initialize(IEventSource eventSource)
+        {
+            Environment.SetEnvironmentVariable("MSBUILDTARGETOUTPUTLOGGING", "true");
```

Comment:
This line shows that out of proc nodes also need to know about this setting: [https://github.com/Microsoft/msbuild/blob/xplat/src/Build/BackEnd/Components/Communications/LogMessagePacket.cs#L63](https://github.com/Microsoft/msbuild/blob/xplat/src/Build/BackEnd/Components/Communications/LogMessagePacket.cs#L63)
To fix this to not use environment variables, we'd have to come up with a way of propagating this information to the nodes. Probably by adding a new property to either BuildParameters or NodeConfiguration.

Another smell is that the environment var is read in two different places in the msbuild codebase. It could be unified and put into LoggingService.


---
In reply to: [103011070](https://github.com/Microsoft/msbuild/pull/1747#discussion_r103011070) [](ancestors = 103011070)